apps/vmq_server/test/vmq_cluster_SUITE.erl

```diff
@@ -543,6 +545,56 @@ shared_subs_random_policy_test(Config) ->
     [ ok = gen_tcp:close(S) || S <- SubscriberSockets ],
     ok.
 
+restarted_node_has_no_stale_sessions(Config) ->
+    ok = ensure_cluster(Config),
+    {_, Nodes} = lists:keyfind(nodes, 1, Config),
+    [{RestartNodeName, RestartNodePort},
+     {_, OtherNodePort}|_] = Nodes,
+
+    %% Connect and subscribe on a node with cs true
+    Topic = <<"topic">>,
+    ClientId = "cs-true-client-id",
+    ClientConn = connect(RestartNodePort, ClientId, [{keepalive, 60}, {clean_session, true}]),
+    subscribe(ClientConn, Topic, 1),
+
+    ToRemain =
+        [#{client_id => "cs-false-client-id",
+           port => RestartNodePort,
+           clean_session => false},
+         #{client_id => "cs-true-false-client-id",
+           port => OtherNodePort,
+           clean_session => false}],
+    [ begin
+          Conn = connect(P, C, [{keepalive, 60}, {clean_session, CS}]),
+          subscribe(Conn, Topic, 1)
+      end || #{client_id := C, port := P, clean_session := CS} <- ToRemain ],
+
+    %% Make sure subscriptions have propagated to all nodes
+    ok = wait_until_converged(Nodes,
+                              fun(N) ->
+                                      rpc:call(N, vmq_reg, total_subscriptions, [])
+                              end, [{total, 3}]),
+
+    %% Restart the node.
+    {ok, _} = rpc:call(RestartNodeName, vmq_server_cmd, node_stop, []),
+    {ok, _} = rpc:call(RestartNodeName, vmq_server_cmd, node_start, []),
+    Sessions = rpc:call(RestartNodeName, vmq_ql_query_mgr, fold_query,
+                        [fun(V, Acc) -> [V|Acc] end, [], "SELECT * FROM sessions"]),
+    {Removed, Remaining} =
+        lists:foldl(
+          fun(#{node := N, client_id := C} = S, {Removed, Other})
+                when N =:= RestartNodeName, C =:= ClientId ->
+                  {[S|Removed], Other};
+             (S, {Removed, Other}) ->
+                  {Removed, [S|Other]}
+          end,
+          {[], []},
+          Sessions),
+
+    [] = Removed,
```

Comment:
Hey @dergraf I rewrote the test to be more directly comparing things... I think (hope) it's clearer now.